### ISSUE TITLE
NullPointerException: Cannot invoke "TypeBinding.kind()" because "type" is null

### ISSUE THREAD

**Issue by liefke** [eclipse-jdt/eclipse.jdt.core#986]
I just migrated from Eclipse 2022-12 to 2023-03.

Now I get the following exception when building or editing some of the classes in my project:

```
java.lang.NullPointerException: Cannot invoke "org.eclipse.jdt.internal.compiler.lookup.TypeBinding.kind()" because "type" is null
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.convertToRawType(LookupEnvironment.java:844)
	at org.eclipse.jdt.internal.compiler.lookup.RawTypeBinding.initializeArguments(RawTypeBinding.java:238)
	at org.eclipse.jdt.internal.compiler.lookup.ParameterizedTypeBinding.substitute(ParameterizedTypeBinding.java:1432)
	at org.eclipse.jdt.internal.compiler.lookup.Scope$Substitutor.substitute(Scope.java:586)
	at org.eclipse.jdt.internal.compiler.lookup.Scope.substitute(Scope.java:530)
	at org.eclipse.jdt.internal.compiler.lookup.ParameterizedFieldBinding.<init>(ParameterizedFieldBinding.java:36)
	at org.eclipse.jdt.internal.compiler.lookup.ParameterizedTypeBinding.fields(ParameterizedTypeBinding.java:609)
	at org.eclipse.jdt.internal.compiler.lookup.ParameterizedTypeBinding.getField(ParameterizedTypeBinding.java:803)
	at org.eclipse.jdt.internal.compiler.lookup.Scope.findField(Scope.java:1437)
	at org.eclipse.jdt.internal.compiler.lookup.Scope.findField(Scope.java:1383)
	at org.eclipse.jdt.internal.compiler.lookup.MethodScope.findField(MethodScope.java:473)
	at org.eclipse.jdt.internal.compiler.lookup.BlockScope.getBinding(BlockScope.java:623)
	at org.eclipse.jdt.internal.compiler.ast.QualifiedNameReference.resolveType(QualifiedNameReference.java:1029)
	at org.eclipse.jdt.internal.compiler.ast.MemberValuePair.resolveTypeExpecting(MemberValuePair.java:93)
	at org.eclipse.jdt.internal.compiler.ast.Annotation.resolveType(Annotation.java:932)
	at org.eclipse.jdt.internal.compiler.ast.ASTNode.resolveAnnotations(ASTNode.java:958)
	at org.eclipse.jdt.internal.compiler.ast.ASTNode.resolveAnnotations(ASTNode.java:788)
	at org.eclipse.jdt.internal.compiler.lookup.SourceTypeBinding.initAnnotationTagBits(SourceTypeBinding.java:1657)
	at org.eclipse.jdt.internal.compiler.lookup.SourceTypeBinding.getAnnotationTagBits(SourceTypeBinding.java:1642)
	at org.eclipse.jdt.internal.compiler.lookup.SourceTypeBinding.getNullDefault(SourceTypeBinding.java:2054)
	at org.eclipse.jdt.internal.compiler.lookup.ClassScope.hasDefaultNullnessFor(ClassScope.java:1737)
	at org.eclipse.jdt.internal.compiler.lookup.ClassScope.findSupertype(ClassScope.java:1671)
	at org.eclipse.jdt.internal.compiler.lookup.ClassScope.connectSuperclass(ClassScope.java:1176)
	at org.eclipse.jdt.internal.compiler.lookup.ClassScope.connectTypeHierarchy(ClassScope.java:1454)
	at org.eclipse.jdt.internal.compiler.lookup.CompilationUnitScope.connectTypeHierarchy(CompilationUnitScope.java:394)
	at org.eclipse.jdt.internal.compiler.lookup.LookupEnvironment.completeTypeBindings(LookupEnvironment.java:521)
	at org.eclipse.jdt.internal.compiler.Compiler.internalBeginToCompile(Compiler.java:881)
	at org.eclipse.jdt.internal.compiler.Compiler.beginToCompile(Compiler.java:394)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:1248)
	at org.eclipse.jdt.core.dom.CompilationUnitResolver.resolve(CompilationUnitResolver.java:790)
	at org.eclipse.jdt.core.dom.ASTParser.internalCreateAST(ASTParser.java:1245)
	at org.eclipse.jdt.core.dom.ASTParser.createAST(ASTParser.java:868)
	at org.eclipse.jdt.core.manipulation.CoreASTProvider$1.run(CoreASTProvider.java:294)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jdt.core.manipulation.CoreASTProvider.createAST(CoreASTProvider.java:286)
	at org.eclipse.jdt.core.manipulation.CoreASTProvider.getAST(CoreASTProvider.java:199)
	at org.eclipse.jdt.core.manipulation.SharedASTProviderCore.getAST(SharedASTProviderCore.java:138)
	at org.eclipse.jdt.internal.ui.viewsupport.SelectionListenerWithASTManager$PartListenerGroup.calculateASTandInform(SelectionListenerWithASTManager.java:166)
	at org.eclipse.jdt.internal.ui.viewsupport.SelectionListenerWithASTManager$PartListenerGroup$1.run(SelectionListenerWithASTManager.java:151)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

I could narrow the problem down to this class:

```
package mypackage;

@Deprecated(since = Example.SINCE)
class Example<T> implements Serializable {
	
	static final String SINCE = "...";

	private T target;

}
```

The real class looks different and is much more complex, `Serializable` and `Deprecated` are just examples. The problem occurs with any other annotation or interface and even when dropping the `implements Serializable` and using a super class instead. If I import `Example` into another class (even without using it), I get the same error for that class, too.

If I remove either the `implements`, the annotation or the `target` attribute, or if I inline the `since` reference, the error is gone. As a workaround I'm using the inline String in the annotation (`@Deprecated(since = "...")`).

**Comment by iloveeclipse** [comment]
I can't reproduce in a new Java project, please attach project with all settings etc.
Note that your example has compile error:
![image](https://user-images.githubusercontent.com/964108/233102623-ca84755e-d581-4351-9fc7-9dea2cf2b326.png)

**Comment by srikanth-sankaran** [comment]
@stephan-herrmann Another problem scenario where org.eclipse.jdt.internal.compiler.lookup.ClassScope.hasDefaultNullnessFor(ClassScope.java:1737) triggers a symptom that is different from https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1007 and cousins

**Comment by stephan-herrmann** [comment]
Will be fixed by the next iteration in #1067

**Comment by stephan-herrmann** [comment]
Resolved via #1077
